### PR TITLE
Add Magic Leap Control Support, and handling for profile absence

### PR DESCRIPTION
**Pull request by atarng-magicleap**
**Description:**
Aframe would not operate smoothly if profiles have not been specified by XR input sources.

**Changes proposed:**
- Added controller support for Magic Leap Devices
- Added fallback handling for if profiles have not been provided.

### CODE REVIEW

File: src/components/magicleap-controls.js [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+ * 2 - grip
+ * 3 - menu (dispatch but better for menu options)
+ *
+ * Axis:
+ * 0 - touchpad x axis
+ * 1 - touchpad y axis
+ */
+var INPUT_MAPPING_WEBVR = {
+  axes: {trackpad: [0, 1]},
+  buttons: ['trigger', 'grip', 'trackpad', 'menu']
```
Comment (yhuang-ml) [inline src/components/magicleap-controls.js]: The buttons order doesn't match the comments above

File: src/components/magicleap-controls.js [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+};
+
+/**
+ * Button IDs:
+ * 0 - trigger
+ * 1 - squeeze
```
Comment (yhuang-ml) [inline src/components/magicleap-controls.js]: squeeze - do you mean grip?

File: src/components/magicleap-controls.js [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+ * 2 - touchpad
```
Comment (yhuang-ml) [inline src/components/magicleap-controls.js]: Should it be touchpad or trackpad?

File: src/components/magicleap-controls.js [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+ * 1 - trigger (intensity value from 0.5 to 1)
+ * 2 - grip
+ * 3 - menu (dispatch but better for menu options)
+ *
+ * Axis:
+ * 0 - touchpad x axis
+ * 1 - touchpad y axis
+ */
+var INPUT_MAPPING_WEBVR = {
+  axes: {trackpad: [0, 1]},
```
Comment (yhuang-ml) [inline src/components/magicleap-controls.js]: Should it be `touchpad` or `trackpad`? The comments above says `touchpad` but here it's `trackpad`.

File: src/components/magicleap-controls.js [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+ * 1 - squeeze
+ * 2 - touchpad
+ * 4 - menu (never dispatched on this layer)
+ *
+ * Axis:
+ * 0 - touchpad x axis
+ * 1 - touchpad y axis
+ */
+var INPUT_MAPPING_WEBXR = {
+  axes: {trackpad: [0, 1]},
```
Comment (yhuang-ml) [inline src/components/magicleap-controls.js]: Should it be `trackpad` or `touchpad`?

File: src/components/tracked-controls-webxr.js [hunk 6]
```diff
@@ -62,17 +62,25 @@ module.exports.Component = registerComponent('tracked-controls-webxr', {
   },
 
   emitButtonDownEvent: function (evt) {
-    if (!this.controller || evt.inputSource.handedness !== this.data.hand) { return; }
+    if (!this.controller) { return; }
     if (this.controller.gamepad) { return; }
+    if (evt.inputSource.handedness !== 'none' &&
```
Comment (dmarcos) [inline src/components/tracked-controls-webxr.js]: This shoud not be necessary or handled somewhere else. Can you explain why this is needed?

File: src/index.js [hunk 7]
```diff
@@ -89,7 +89,7 @@ require('./core/a-mixin');
 require('./extras/components/');
 require('./extras/primitives/');
 
-console.log('A-Frame Version: 1.0.3 (Date 2020-01-15, Commit #746be947)');
```
Comment (dmarcos) [inline src/index.js]: Omit this from PR.

File: src/systems/tracked-controls-webxr.js [hunk 8]
```diff
@@ -19,7 +19,16 @@ module.exports.System = registerSystem('tracked-controls-webxr', {
...
     var xrSession = this.el.xrSession;
     var self = this;
-    if (!xrSession) { return; }
+    if (!xrSession) {
+      // Broadcast that we now have zero controllers connected if there is
+      // no session
+      if (this.oldControllersLength > 0) {
+        this.oldControllersLength = 0;
+        this.controllers = [];
+        this.el.emit('controllersupdated', undefined, false);
```
Comment (dmarcos) [inline src/systems/tracked-controls-webxr.js]: Why is this necessary? `controllerdisconnected` should be [fired](https://github.com/aframevr/aframe/blob/master/src/utils/tracked-controls.js#L44)

Comment (atarng-magicleap) [reply src/systems/tracked-controls-webxr.js]: From what I'm seeing, we hit an error which states that we have "Failed to execute 'getPose' on 'XRFrame': XRSpace and XRFrame sessions do not match.", which seems to stem from an old session id provided by the controllers.  This appears to be cached from the system.controllers.

File: src/components/magicleap-controls.js (hunk 4 above)
Comment (dmarcos) [reply src/components/magicleap-controls.js]: You can omit unless there’s any Magic Leap browser that is shipping WebVR

File: src/components/magicleap-controls.js (hunk 2 above)
Comment (dmarcos) [reply src/components/magicleap-controls.js]: Yes squeeze == grip usually. That’s WebXR jargon

File: src/components/magicleap-controls.js (hunk 5 above)
Comment (dmarcos) [reply src/components/magicleap-controls.js]: Trackpad. Adjust comment

File: src/components/magicleap-controls.js [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+  buttons: ['trigger', 'grip', 'trackpad', 'menu']
+};
+
+var INPUT_MAPPING = isWebXRAvailable ? INPUT_MAPPING_WEBXR : INPUT_MAPPING_WEBVR;
```
Comment (dmarcos) [inline src/components/magicleap-controls.js]: We can omit all WebVR mappings since I don’t think there’s any browser shipping the old API on Magic Leap

**Comment by dmarcos** [comment]
@atarng-magicleap Let me know if you have any questions or need any help. Just noticed that the PR was closed.
